### PR TITLE
Add special mode handling for tuples in matches and let bindings

### PR DESCRIPTION
**Pull request by lpw25**
Adds special handing of modes when matching directly on a tuple. For example, the following is now allowed:
```ocaml
let escape : string -> unit = fun x -> ()

let foo (local_ x) y =
  match x, y with
  | Some _, Some b -> escape b
  | None, _ -> ()
  | _  -> ();;
```
whilst this will still fail:
```ocaml
let foo (local_ x) y =
  let pr = x, y in
  match pr with
  | Some _, Some b -> escape b
  | None, _ -> ()
  | _  -> ();;
```
This PR is on top of #37, #35 and #34.

### CODE REVIEW

**Comment by stedolan** [comment]
As part of this PR, would you mind also squashing the FIXME in `Matching.do_for_multiple_match`: if there is a match on a tuple, and the whole tuple is actually used, then the tuple is allocated in mode `Alloc_heap`, which is not always correct. Here's an example that crashes in the debug runtime:
```ocaml
let[@inline never] f _ = ()

let[@inline never] g (local_ a) (local_ b) =
  match a, b with
  | z -> f z; ()

let () =
  g {contents=1} 2;
  Gc.minor ()
```

(I left a FIXME here a while ago since it at the time it was awkward to work out the correct mode to use. I think that might be easier with this patch)

**Comment by lpw25** [comment]
I've added a commit that fixes that issue.

I've also relaxed the rules slightly so that this is also allowed:
```ocaml
let foo (local_ x) y =
  match x, y with
  | Some _, Some b -> escape b
  | None, _ -> ()
  | pr  -> let _, _ = pr in ();;
```
previously the presence of a match on the whole tuple (`pr`) was disabling the special handling.

**Comment by stedolan** [comment]
I haven't read the code fully yet, but a remark: the approach is more complicated than I was expecting, extending the `expected_mode` of each term with a tuple mode as well. (I was expecting a syntactic match on `Pexp_tuple` as the bound expression of a `let` or the scrutinee of a `match`, rather than plumbing tuple modes around).

As far as I understand, the major feature that the more complicated approach brings is the ability to track tuple modes through deeply nested expressions, e.g. independently tracking modes of `a` and `b` in:
```ocaml
let a, b =
  match ... with
  | None -> p, q
  | Some _ -> r, s
```
(This seems useful enough to be worth the extra complexity, but it's not what I was expecting)

File: typing/typecore.ml [hunk 1]
```diff
@@ -2006,13 +2053,14 @@ and type_pat_aux
                             loc, sp.ppat_attributes) :: p.pat_extra }
       )
   | Ppat_exception p ->
-      type_pat Value p Predef.type_exn (fun p_exn ->
+      let alloc_mode = simple_pat_mode Value_mode.global in
+      type_pat Value ~alloc_mode p Predef.type_exn (fun p_exn ->
```
Comment (stedolan) [inline typing/typecore.ml:2059]: This looks like an unrelated but good change that improves the typing of match-with-exception - it should probably have an associated test.

File: typing/typecore.ml [hunk 2]
```diff
@@ -210,60 +210,79 @@ type tail_or_nontail = Tail | Nontail
 type expected_mode =
   { position : tail_or_nontail;
     escaping_context : escaping_context option;
-    mode : Value_mode.t }
+    mode : Value_mode.t;
+    tuple_modes : Value_mode.t list; }
```
Comment (stedolan) [inline typing/typecore.ml]: I think there's an invariant that for all `t` in `tuple_modes`, `t <= mode` (which would be worth a comment)

File: typing/typecore.ml [hunk 3]
```diff
@@ -210,60 +210,79 @@ type tail_or_nontail = Tail | Nontail
...
+
+let mode_tuple mode tuple_modes =
+  let position = Nontail in
+  let escaping_context = None in
+  { position; escaping_context; mode; tuple_modes }
 
 let submode ~loc ~env mode expected_mode =
-  match Types.Value_mode.submode mode expected_mode.mode with
+  let modes = expected_mode.mode :: expected_mode.tuple_modes in
+  match Types.Value_mode.submode_meet mode modes  with
```
Comment (stedolan) [inline typing/typecore.ml]: Due to the invariant mentioned above, this does a redundant check when there are tuple modes. Consider instead:
```ocaml
match expected_mode.tuple_modes with
| [] -> Types.Value_mode.submode mode expected_mode.mode
| tuples -> Types.Value_mode.submode_meet mode tuples
```

**Review by stedolan — APPROVED** [review]
This patch is more elaborate than I was expecting (I probably wouldn't have bothered supporting mode-tracking for `let (x, y) as p = ...`, for instance), but it's not a huge amount of code and looks correct so lgtm!

File: typing/typecore.ml (hunk 2 above)
Comment (lpw25) [reply typing/typecore.ml]: It's actually the slightly stronger `regional_to_local t <= mode` (or equivalently `t <= regional_to_global mode`). I'll add a comment.

**Pull request merged**